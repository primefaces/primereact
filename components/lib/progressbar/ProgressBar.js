import React, { forwardRef, memo } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/Utils';

export const ProgressBar = memo(forwardRef((props, ref) => {

    const createLabel = () => {
        if (props.showValue && props.value != null) {
            const label = props.displayValueTemplate ? props.displayValueTemplate(props.value) : props.value + props.unit;
            return <div className="p-progressbar-label">{label}</div>
        }

        return null;
    }

    const createDeterminate = () => {
        const className = classNames('p-progressbar p-component p-progressbar-determinate', props.className);
        const label = createLabel();

        return (
            <div role="progressbar" id={props.id} className={className} style={props.style} aria-valuemin="0" aria-valuenow={props.value} aria-valuemax="100" aria-label={props.value}>
                <div className="p-progressbar-value p-progressbar-value-animate" style={{ width: props.value + '%', display: 'block', backgroundColor: props.color }}></div>
                {label}
            </div>
        )
    }

    const createIndeterminate = () => {
        const className = classNames('p-progressbar p-component p-progressbar-indeterminate', props.className);

        return (
            <div role="progressbar" id={props.id} className={className} style={props.style}>
                <div className="p-progressbar-indeterminate-container">
                    <div className="p-progressbar-value p-progressbar-value-animate" style={{ backgroundColor: props.color }}></div>
                </div>
            </div>
        )
    }

    if (props.mode === 'determinate')
        return createDeterminate();
    else if (props.mode === 'indeterminate')
        return createIndeterminate();
    else
        throw new Error(props.mode + " is not a valid mode for the ProgressBar. Valid values are 'determinate' and 'indeterminate'");
}));

ProgressBar.defaultProps = {
    __TYPE: 'ProgressBar',
    id: null,
    value: null,
    showValue: true,
    unit: '%',
    style: null,
    className: null,
    mode: 'determinate',
    displayValueTemplate: null,
    color: null
}

ProgressBar.propTypes /* remove-proptypes */ = {
    __TYPE: PropTypes.string,
    id: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    showValue: PropTypes.bool,
    unit: PropTypes.string,
    style: PropTypes.object,
    className: PropTypes.string,
    mode: PropTypes.string,
    displayValueTemplate: PropTypes.func,
    color: PropTypes.string
}
