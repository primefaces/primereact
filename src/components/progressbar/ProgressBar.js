import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/Utils';

export class ProgressBar extends Component {

    static defaultProps = {
        id: null,
        value: null,
        showValue: true,
        unit: '%',
        style: null,
        className: null,
        mode: 'determinate',
        displayValueTemplate: null,
        color: null
    }

    static propTypes = {
        id: PropTypes.string,
        value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        showValue: PropTypes.bool,
        unit: PropTypes.string,
        style: PropTypes.object,
        className: PropTypes.string,
        mode: PropTypes.string,
        displayValueTemplate: PropTypes.func,
        color: PropTypes.string
    };

    renderLabel() {
        if (this.props.showValue && this.props.value != null) {
            let label = this.props.displayValueTemplate ? this.props.displayValueTemplate(this.props.value) : this.props.value + this.props.unit;
            return (
                <div className="p-progressbar-label">{label}</div>
            );
        }

        return null;
    }

    renderDeterminate() {
        let className = classNames('p-progressbar p-component p-progressbar-determinate', this.props.className);
        let label = this.renderLabel();

        return (
            <div role="progressbar" id={this.props.id} className={className} style={this.props.style} aria-valuemin="0" aria-valuenow={this.props.value} aria-valuemax="100" aria-label={this.props.value}>
                <div className="p-progressbar-value p-progressbar-value-animate" style={{ width: this.props.value + '%', display: 'block', backgroundColor: this.props.color }}></div>
                {label}
            </div>
        );
    }

    renderIndeterminate() {
        let className = classNames('p-progressbar p-component p-progressbar-indeterminate', this.props.className);

        return (
            <div role="progressbar" id={this.props.id} className={className} style={this.props.style}>
                <div className="p-progressbar-indeterminate-container">
                    <div className="p-progressbar-value p-progressbar-value-animate" style={{ backgroundColor: this.props.color }}></div>
                </div>
            </div>
        )
    }

    render() {
        if (this.props.mode === 'determinate')
            return this.renderDeterminate();
        else if (this.props.mode === 'indeterminate')
            return this.renderIndeterminate();
        else
            throw new Error(this.props.mode + " is not a valid mode for the ProgressBar. Valid values are 'determinate' and 'indeterminate'");
    }

}
