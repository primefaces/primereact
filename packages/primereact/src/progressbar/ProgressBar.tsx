'use client';
import { Component, withComponent } from '@primereact/core/component';
import { useProgressBar } from '@primereact/headless/progressbar';
import { styles } from '@primereact/styles/progressbar';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { defaultProps } from './ProgressBar.props';
import { ProgressBarLabel } from './label';

export const ProgressBar = withComponent({
    defaultProps,
    styles,
    setup: (instance) => {
        const progressBar = useProgressBar(instance.inProps);

        return progressBar;
    },
    render: (instance) => {
        const {
            id,
            props,
            ptmi,
            ptm,
            cx,
            // element refs
            elementRef
        } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root'),
                style: props.style,
                role: 'progressbar',
                'aria-valuemin': 0,
                'aria-valuenow': props.value,
                'aria-valuemax': 100
            },
            ptmi('root')
        );

        const createDeterminate = () => {
            const valueProps = mergeProps(
                {
                    className: cx('value'),
                    style: {
                        width: props.value + '%',
                        display: 'flex'
                    }
                },
                ptm('value')
            );

            return <div {...valueProps}>{props.children}</div>;
        };

        const createIndeterminate = () => {
            const valueProps = mergeProps(
                {
                    className: cx('value')
                },
                ptm('value')
            );

            return <div {...valueProps}></div>;
        };

        return (
            <Component as={props.as || 'div'} {...rootProps} ref={elementRef}>
                {props.mode === 'determinate' ? createDeterminate() : createIndeterminate()}
            </Component>
        );
    },
    components: {
        Label: ProgressBarLabel
    }
});
