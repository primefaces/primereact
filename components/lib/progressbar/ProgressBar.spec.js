import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { snapshot } from '../../test';
import { ProgressBar } from './ProgressBar';

describe('ProgressBar', () => {
    snapshot(<ProgressBar />, 'default');
    snapshot(<ProgressBar value={50} />, 'value');
    snapshot(<ProgressBar value={50} showValue={false} />, 'hide value');
    snapshot(<ProgressBar value={50} unit="%" />, 'value with unit');
    snapshot(<ProgressBar mode="determinate" value={20} />, 'mode determinate');
    snapshot(<ProgressBar mode="indeterminate" style={{ height: '6px' }} />, 'mode indeterminate');
    test('when mode is invalid throws exception', () => {
        try {
            const { container } = render(<ProgressBar mode="invalid" value={20} />);
        } catch (error) {
            expect(error.toString().startsWith("Error: invalid is not a valid mode for the ProgressBar. Valid values are 'determinate' and 'indeterminate'")).toBeTruthy();
        }
    });
});
