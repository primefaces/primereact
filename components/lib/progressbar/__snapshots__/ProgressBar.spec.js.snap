// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar default 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    class="p-progressbar p-component p-progressbar-determinate"
    role="progressbar"
  >
    <div
      class="p-progressbar-value p-progressbar-value-animate"
      style="display: flex;"
    />
  </div>
</div>
`;

exports[`ProgressBar hide value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="p-progressbar p-component p-progressbar-determinate"
    role="progressbar"
  >
    <div
      class="p-progressbar-value p-progressbar-value-animate"
      style="width: 50%; display: flex;"
    />
  </div>
</div>
`;

exports[`ProgressBar mode determinate 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="20"
    class="p-progressbar p-component p-progressbar-determinate"
    role="progressbar"
  >
    <div
      class="p-progressbar-value p-progressbar-value-animate"
      style="width: 20%; display: flex;"
    >
      <div
        class="p-progressbar-label"
      >
        20%
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressBar mode indeterminate 1`] = `
<div>
  <div
    class="p-progressbar p-component p-progressbar-indeterminate"
    role="progressbar"
    style="height: 6px;"
  >
    <div
      class="p-progressbar-indeterminate-container"
    >
      <div
        class="p-progressbar-value p-progressbar-value-animate"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressBar value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="p-progressbar p-component p-progressbar-determinate"
    role="progressbar"
  >
    <div
      class="p-progressbar-value p-progressbar-value-animate"
      style="width: 50%; display: flex;"
    >
      <div
        class="p-progressbar-label"
      >
        50%
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressBar value with unit 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="p-progressbar p-component p-progressbar-determinate"
    role="progressbar"
  >
    <div
      class="p-progressbar-value p-progressbar-value-animate"
      style="width: 50%; display: flex;"
    >
      <div
        class="p-progressbar-label"
      >
        50%
      </div>
    </div>
  </div>
</div>
`;
